YNU-354: Protocol Docs (#69)
layer-3/docs: docs/protocol/on-chain/data-structures.mdx

:::info Participant Addresses
The caller (EOA or contract) is usually different from the <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participant</Tooltip> address. Participants are derived addresses generated by the owner, similar to <Tooltip content="A temporary cryptographic key delegated by a user's main wallet that provides a flexible way for the user to manage security of their funds by giving specific permissions and allowances for specific apps.">session keys</Tooltip>, to enable high-frequency state signing without exposing the main wallet.
:::

## State

Represents a snapshot of <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">channel</Tooltip> state at a point in time.

```solidity
struct State {
    bytes data;               // Application-specific data
    Allocation[] allocations; // Asset allocation for each participant

>     Signature[] sigs;         // Signatures from participants

Reviewer: It also has `version` and `intent` fields

Author: Missed it, added it right away

}
```

**Fields**:

- `data`: Application-specific data. For funding states, contains magic numbers (CHANOPEN = 7877, CHANCLOSE = 7879).
- `allocations`: Array of allocations defining how funds are distributed.
- `sigs`: Array of signatures, one per <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participant</Tooltip>. Order corresponds to participants array in Channel.

## Allocation

Specifies how a particular amount of a token should be allocated.

```solidity
struct Allocation {
    address destination;  // Recipient of funds
    address token;        // ERC-20 token address
    uint256 amount;       // Token amount in smallest unit
}
```

**Fields**:

- `destination`: Address that will receive the funds when channel closes.
- `token`: Contract address of the ERC-20 token (or zero address for native currency).
- `amount`: Amount in the token's smallest unit (wei for ETH, considering decimals for ERC-20).

## Signature

> ECDSA signature components.

Reviewer: Actually, we're not restricted to ECDSA signatures, but I think @nksazonov can elaborate better here

Author: I was able to find only mention of ECDSA, please let me know this is not the case

Reviewer 2: Indeed, we are not restricted to ECDS algorithm signatures.
Firstly, the output of ECDSA is a bunch numbers, which can be represented in several popular formats: `r,s,v` and 65 bytes.
Previously we indeed used the `r,s,v`, but since v0.3.0 release we have switched to `bytes` to be able to support non-ECDSA signatures, e.g. ERC1271 and ERC6492 ones.

Author: Noted

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

    address token;    // ERC-20 token address
    uint256 amount;   // Token amount
}
```

## Channel Status

Enum representing the lifecycle stage of a channel.

```solidity
enum Status {
    VOID,      // Channel does not exist
    INITIAL,   // Creation in progress, awaiting all participants
    ACTIVE,    // Fully funded and operational
    DISPUTE,   // Challenge period active
    FINAL      // Ready to be closed and deleted
}
```

**State Transitions**:

```mermaid
stateDiagram-v2
    [*] --> VOID
    VOID --> INITIAL: create()
    INITIAL --> ACTIVE: join() (all participants)
    ACTIVE --> DISPUTE: challenge()
    ACTIVE --> FINAL: close() (cooperative)
    DISPUTE --> ACTIVE: checkpoint() (newer state)
    DISPUTE --> FINAL: challenge period expires
    FINAL --> [*]
    
    note right of ACTIVE: Operational state<br/>Off-chain updates occur here
    note right of DISPUTE: Challenge period active<br/>Parties can submit newer states
```

## Protocol Constants

```solidity
constant uint256 CREATOR = 0;      // Creator's participant index
constant uint256 BROKER = 1;       // Broker's participant index
constant uint32 CHANOPEN = 7877;   // Magic number for funding state

> constant uint32 CHANCLOSE = 7879;  // Magic number for closing state

Reviewer: I'm not sure about following constants, @nksazonov please elaborate

Author: I updated these after referring to interfaces/Types.sol.

```javascript
channelId = keccak256(
    abi.encode(
        channel.participants,
        channel.adjudicator,
        channel.challenge,
        channel.nonce
    )
)
```

This creates a deterministic, unique identifier for each <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">channel</Tooltip>. <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain.">App sessions</Tooltip> use the same computation method.

:::note Deterministic IDs
Channel IDs are deterministically computed from the channel configuration, ensuring the same configuration always produces the same identifier.
:::

### State Hash

The <Tooltip content="A cryptographic hash of a channel state, used for signature verification.">stateHash</Tooltip> MUST be computed as:

```javascript
stateHash = keccak256(
    abi.encode(
        channelId,
        state.data,

>         state.allocations

Reviewer: Also includes `intent` and `version`

Author: Missed that. Thanks

```solidity
struct Channel {
    address[] participants;  // List of participants in the channel
    address adjudicator;     // Contract that validates state transitions
    uint64 challenge;        // Duration in seconds for dispute resolution
    uint64 nonce;           // Unique identifier for the channel
}
```

**Fields**:

- `participants`: An ordered array of <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participant</Tooltip> addresses. Index 0 is typically the <Tooltip content="Most often a light client willing to fund a ledger account with a clearnode.">Creator</Tooltip>, index 1 is the <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.">clearnode</Tooltip>.
- `adjudicator`: Address of the <Tooltip content="A smart contract that validates state transitions according to application-specific rules.">adjudicator</Tooltip> contract responsible for validating state transitions.

> - `challenge`: Challenge period duration in seconds. Determines how long a party has to respond to a challenge.

Reviewer 2: ```suggestion
- `challenge`: Challenge period duration in seconds. Determines a time window when a challenge can be resolved by a counterparty. Otherwise, a channel is considered closed and funds can be withdrawn.
```

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

- `nonce`: A unique number that, combined with other fields, creates a unique <Tooltip content="A unique identifier for a channel, computed as the hash of the channel configuration.">channel identifier</Tooltip>.

:::info Participant Addresses

> The caller (EOA or contract) is usually different from the <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participant</Tooltip> address. Participants are derived addresses generated by the owner, similar to <Tooltip content="A temporary cryptographic key delegated by a user's main wallet that provides a flexible way for the user to manage security of their funds by giving specific permissions and allowances for specific apps.">session keys</Tooltip>, to enable high-frequency state signing without exposing the main wallet.

Reviewer 2: Actually, this will become obsolete in v0.5.0 release. First participant will always be user's wallet address.
Thus, I suggest to describe the "Participant vs Caller" difference

```suggestion
:::info Participant versus Caller address
The first <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participant</Tooltip> address  is usually different from the caller (EOA or contract), thus enabling channel operation delegation. This can be fruitful as users can fund channels for other ones.
```

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

:::

## State

Represents a snapshot of <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">channel</Tooltip> state at a point in time.

```solidity
struct State {
    StateIntent intent;       // Intent of the state (INITIALIZE, OPERATE, RESIZE, FINALIZE)
    uint256 version;          // State version incremental number to compare most recent
    bytes data;               // Application-specific data
    Allocation[] allocations; // Asset allocation for each participant
    bytes[] sigs;             // stateHash signatures from participants
}
```

**Fields**:

- `intent`: The intent of this state, indicating its purpose (see <Tooltip content="An enum defining the purpose of a state: INITIALIZE (0), OPERATE (1), RESIZE (2), FINALIZE (3)">StateIntent</Tooltip> enum).
- `version`: Incremental version number used to compare and validate state freshness. Higher versions supersede lower versions.

> - `data`: Application-specific data. For funding states, contains magic numbers (CHANOPEN = 7877, CHANCLOSE = 7879).

Reviewer 2: This is outdated. Magic numbers were replaced by `StateIntent`. However, in case of `resize(...)` `data` indeed contains some additional information:

```suggestion
- `data`: Application-specific data which adjudicators can operate on. For a `resize(...)` state must contain `allocationDeltas`. For more information, please check the [resize operation docs](./channel-lifecycle#resize-protocol).
```

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

- `allocations`: Array of allocations defining how funds are distributed.
- `sigs`: Array of signatures, one per <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participant</Tooltip>. Order corresponds to participants array in Channel.

## Allocation

Specifies how a particular amount of a token should be allocated.

```solidity
struct Allocation {
    address destination;  // Recipient of funds
    address token;        // ERC-20 token address
    uint256 amount;       // Token amount in smallest unit
}
```

**Fields**:

- `destination`: Address that will receive the funds when channel closes.
- `token`: Contract address of the ERC-20 token (or zero address for native currency).
- `amount`: Amount in the token's smallest unit (wei for ETH, considering decimals for ERC-20).

> ## Signature

Reviewer 2: As our current signature support is vast, I suggest creating a separate "Signature formats" page that will explain this in details.
Here however, we can still mention that signature is represented as `bytes` and we support several formats:
- pre-EIP-191 / raw ECDSA signatures, where a raw state hash is signed
- EIP-191 version `0x45` (Ethereum signed message) signatures, which are composed by adding a corresponding string and a message length to a packed state
- EIP-712 signatures, which are composed with a specific `domainSeparator` and a `hasStruct(...)` over a packed state
- EIP-1271 (smart contract) signatures, which can contain arbitrary data, depending on the validator smart contract
- EIP-6492 (predeploy Smart Wallet) signatures, which sole purpose is to validate that a not-yet-deployed ERC-4337 Smart Wallet has signed a message.

@ihsraham (and everyone interested), we can have a call next week where I can explain all supported signature formats in detail.

Author: Sign me up for it!

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

Added a draft page for it, will update it or enhance it post our session !

ECDSA signature components.

```solidity
struct Signature {
    uint8 v;      // Recovery identifier
    bytes32 r;    // First 32 bytes of signature
    bytes32 s;    // Second 32 bytes of signature
}
```

Signatures follow the secp256k1 curve standard used in Ethereum. The protocol is chain-agnostic and signatures are computed over the raw stateHash without EIP-191 prefix.

:::caution Chain-Agnostic Signatures
Signatures in Nitrolite are computed over raw stateHash without EIP-191 prefix, ensuring chain-agnostic compatibility.
:::

## Amount

Represents a quantity of a specific token.

```solidity
struct Amount {
    address token;    // ERC-20 token address
    uint256 amount;   // Token amount
}
```

## Channel Status

Enum representing the lifecycle stage of a channel.

```solidity
enum Status {
    VOID,      // Channel does not exist
    INITIAL,   // Creation in progress, awaiting all participants
    ACTIVE,    // Fully funded and operational
    DISPUTE,   // Challenge period active
    FINAL      // Ready to be closed and deleted
}
```

> **State Transitions**:

Reviewer 2: I think it is better to move this diagram to "channel-lifecycle" page and just mention it here.

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

Reviewer 2: The diagram is still present in the `data-structures.mdx` file.

```mermaid
stateDiagram-v2
    [*] --> VOID
    VOID --> INITIAL: create()
    INITIAL --> ACTIVE: join() (all participants)
    ACTIVE --> DISPUTE: challenge()
    ACTIVE --> FINAL: close() (cooperative)
    DISPUTE --> ACTIVE: checkpoint() (newer state)
    DISPUTE --> FINAL: challenge period expires
    FINAL --> [*]
    
    note right of ACTIVE: Operational state<br/>Off-chain updates occur here
    note right of DISPUTE: Challenge period active<br/>Parties can submit newer states
```

## Protocol Constants

```solidity
constant uint256 CLIENT_IDX = 0;   // Client/Creator participant index
constant uint256 SERVER_IDX = 1;   // Server/Clearnode participant index
constant uint256 PART_NUM = 2;     // Number of participants (always 2)

constant uint16 CHANOPEN = 7877;   // 0x1EC5 - Magic number for funding state

> constant uint16 CHANCLOSE = 7879;  // 0x1EC7 - Magic number for closing state

Reviewer 2: As said before, magic numbers are deprecated.
Instead, Custody defines the following additional constants:

```sol
uint256 public constant MIN_CHALLENGE_PERIOD = 1 hours;

bytes32 public constant CHALLENGE_STATE_TYPEHASH = keccak256(
        "AllowChallengeStateHash(bytes32 channelId,uint8 intent,uint256 version,bytes data,Allocation[] allocations)Allocation(address destination,address token,uint256 amount)"
    );
```

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

```

:::tip Magic Numbers
- **CHANOPEN (7877 / 0x1EC5)**: Signals an initial funding state during channel creation
- **CHANCLOSE (7879 / 0x1EC7)**: Signals a final closing state for cooperative closure
:::

## Identifier Computation

### Channel Identifier

The <Tooltip content="A unique identifier for a channel, computed as the hash of the channel configuration.">channelId</Tooltip> MUST be computed as:

```javascript

> channelId = keccak256(
>     abi.encode(
>         channel.participants,
>         channel.adjudicator,
>         channel.challenge,
>         channel.nonce
>     )

Reviewer 2: ```suggestion
    abi.encode(
        channel.participants,
        channel.adjudicator,
        channel.challenge,
        channel.nonce,
        chainId
    )
```

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

)
```

> This creates a deterministic, unique identifier for each <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">channel</Tooltip>. <Tooltip content="Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain.">App sessions</Tooltip> use the same computation method.

Reviewer 2: As on-chain channels use `chainId` to compute channelIds, the app sessions can not do the same.
@dimast-x  or @philanton , please advice here.

Reviewer: Yeah, the algo is similar, but different

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

:::note Deterministic IDs
Channel IDs are deterministically computed from the channel configuration, ensuring the same configuration always produces the same identifier.
:::

> ### State Hash

Reviewer 2: The "state hash" concept was deprecated in v0.3.0 with introduction of non-ECDSA signatures support. Instead, an "packed state" concept should be used.
It is computed as: `abi.encode(channelId, state.intent, state.version, state.data, state.allocations);`, so basically a "state hash" without actual hashing.
It is passed to a chosen signing method.

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

The <Tooltip content="A cryptographic hash of a channel state, used for signature verification.">stateHash</Tooltip> MUST be computed as:

```javascript
stateHash = keccak256(
    abi.encode(
        channelId,
        state.intent,
        state.version,
        state.data,
        state.allocations
    )
)
```

Signatures are generated over this <Tooltip content="A cryptographic hash of a channel state, used for signature verification.">stateHash</Tooltip>. The stateHash is signed without EIP-191 prefix as the protocol is intended to be chain-agnostic.

> **Computation Process**:

Reviewer 2: I think this can be safely replaced with just `abi.encode(channelId, state.intent, state.version, state.data, state.allocations);`.

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

```mermaid
graph LR
    A[channelId] --> D[Combine]
    B[state.data] --> D
    C[state.allocations] --> D
    D --> E[abi.encode]
    E --> F[keccak256]
    F --> G[stateHash]
    
    style G fill:#90EE90
```

:::warning Signature Verification

> All state updates MUST be verified by checking signatures against the computed stateHash before accepting them on-chain.

Reviewer 2: ```suggestion
All state updates MUST be verified by checking signatures against the computed `packedState` before accepting them on-chain.
```

Author: [f889dd9d34e1744c4d0d2598c3c590f87acb676b](https://github.com/layer-3/docs/pull/69/commits/f889dd9d34e1744c4d0d2598c3c590f87acb676b)

:::info Participant versus Caller Address
The first <Tooltip content={tooltipDefinitions.participant}>participant</Tooltip> address is usually different from the caller (EOA or contract), thus enabling channel operation delegation. This can be fruitful as users can fund channels for other ones.
:::

## State

Represents a snapshot of <Tooltip content={tooltipDefinitions.channel}>channel</Tooltip> state at a point in time.

```solidity
struct State {
    StateIntent intent;       // Intent of the state (INITIALIZE, OPERATE, RESIZE, FINALIZE)
    uint256 version;          // State version incremental number to compare most recent
    bytes data;               // Application-specific data
    Allocation[] allocations; // Asset allocation for each participant
    bytes[] sigs;             // Participant signatures authorizing the packed state payload
}
```

**Fields**:

- `intent`: The intent of this state, indicating its purpose (see <Tooltip content={tooltipDefinitions.stateIntent}>StateIntent</Tooltip> enum).
- `version`: Incremental version number used to compare and validate state freshness. Higher versions supersede lower versions.

> - `data`: Application-specific data which adjudicators can operate on. For a `resize(...)` state must contain `allocationDeltas`. For more information, please check the [resize operation docs](./channel-lifecycle#resize-protocol).

Reviewer 2: For some strange reason, the `[resize operation docs](./channel-lifecycle#resize-protocol).` link redirects me to the correct (`channel-lifecycle`) page, but to the "Checkpointing" section. Is it the same for you?

Author: Yes, it does, strange.

Author: I did some digging
This is an issue
What's happening is that images and mermaid diagrams load/render later than the text, by that time `element.scrollIntoView()` called takes you to the resize section before loading of images/mermaid diagrams based purely on already loaded text content. Then eventually images/diagrams load and you end up elsewhere.
I believe images with width and height specified might have fixed placeholders set and they would take us to the right section everytime.
I hope I made sense

- `token`: Contract address of the ERC-20 token (or zero address for native currency).
- `amount`: Amount in the token's smallest unit (wei for ETH, considering decimals for ERC-20).

## Signatures

> Signatures in <Tooltip content={tooltipDefinitions.nitroliteProtocol}>Nitrolite</Tooltip> are stored as raw `bytes` so the protocol can validate multiple scheme formats. Each signature ultimately resolves to the tuple `(v, r, s)` used by ECDSA-compatible verifiers, but additional metadata can be embedded in the byte payload when required by the signing standard.

Reviewer 2: The part about "each signature ultimately resolving to a tuple `(v, r, s)`" is not true.

```suggestion
Signatures in <Tooltip content={tooltipDefinitions.nitroliteProtocol}>Nitrolite</Tooltip> are stored as raw `bytes` so the protocol can validate multiple scheme formats.
```

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

## Identifier Computation

### Channel Identifier

The <Tooltip content={tooltipDefinitions.channelId}>channelId</Tooltip> MUST be computed as:

```javascript
channelId = keccak256(
    abi.encode(
        channel.participants,
        channel.adjudicator,
        channel.challenge,
        channel.nonce,
        chainId
    )
)
```

This creates a deterministic, unique identifier for each <Tooltip content={tooltipDefinitions.channel}>channel</Tooltip>.

:::info App Session Identifiers
<Tooltip content={tooltipDefinitions.appChannel}>App sessions</Tooltip> use a different computation: `keccak256(JSON.stringify(definition))` where definition includes the app configuration but **not** `chainId`, since sessions are entirely off-chain. See [Off-chain › App Sessions › Session Identifier](../off-chain/app-sessions#session-identifier) for details.
:::

:::note Deterministic IDs
Channel IDs are deterministically computed from the channel configuration, ensuring the same configuration always produces the same identifier.
:::

### Packed State

> The legacy <Tooltip content={tooltipDefinitions.stateHash}>state hash</Tooltip> concept was removed in v0.3.0 when non-ECDSA signatures were introduced. Instead, participants sign the **packed state** payload directly:

Reviewer 2: In fact, the packed state payload is not directly used for signing, but is a base for all supported signature standards.

```suggestion
The legacy <Tooltip content={tooltipDefinitions.stateHash}>state hash</Tooltip> concept was removed in v0.3.0 when non-ECDSA signatures were introduced. Instead, participants use the **packed state** payload for signing:
```

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)
